Setups code formatters, and polishes scaffold output

smol-ninja opened the pull request:
### This PR addresses the following issues
- Closes https://github.com/alexfertel/bulloak/issues/103
- Closes https://github.com/alexfertel/bulloak/issues/104

### PR dependency

- https://github.com/alexfertel/bulloak/pull/102

### Added

- Adds `.editorconfig`.
- Setups `husky` and `lint-staged` to format automatically when committing changes.
- Adds `prettier` for formatting `json,md,tsx,yml` files.
- Adds `forge` and `solhint` to format solidity files.
- Updates contributing guidelines.
- Formats code according to above rules.
- Added `package.json` to use prettier and other npm packages. Its using multiple common license standard from [npm docs](https://docs.npmjs.com/cli/v11/configuring-npm/package-json).

### Changed
- When `--vm.skip` is provided, `scaffold` will generate `import { Test }` instead of `import {Test}`, a better formatted version.
- Replaced `forge-std/Test.sol` with `forge-std/src/Test.sol` as I am using npm package to setup forge for testing. Let me know if thats OK. Ideally this should somehow be a part of bulloak config where based on user setup, it either generates `forge-std/src/Test.sol` or `forge-std/Test.sol`. Additionally, we, at Sablier, use npm package to install `forge`.
- Instead of printing `// it <description>`, it will now print `// It <description>.` notice capitalized `I` and `.` at the end of it.

alexfertel commented:
Hey! Thank you very much for taking the time!

I see that the scope of the PR is much bigger than I anticipated. There are a few changes that I think will need to be reverted before we can merge this.

> Adds forge and solhint to format solidity files.

This would actually need to be a separate more robust feature, where we read the foundry.toml of the user and read the formatting rules from there.

We want to have differently formatted `.sol` files to double-check `bulloak`'s behaviour. I think we want to revert this change.

> When --vm.skip is provided, scaffold will generate import { Test } instead of import {Test}, a better formatted version.

This is also part of the above feature, and we shouldn't enforce it on the user (or contributor). I've seen many codebases using the current form of imports. [Example 1](https://github.com/0xProject/0x-settler/blob/master/src/core/BalancerV3.sol), [Example 2](https://github.com/Uniswap/v4-core/blob/main/src/PoolManager.sol).

I think we should also revert this change.

> Replaced forge-std/Test.sol with forge-std/src/Test.sol as I am using npm package to setup forge for testing. Let me know if thats OK. Ideally this should somehow be a part of bulloak config where based on user setup, it either generates forge-std/src/Test.sol or forge-std/Test.sol. Additionally, we, at Sablier, also use npm package to install forge.

Yeah, exactly. `npm` is not the industry-standard so I think we shouldn't target it. 

I am open to this specific change, but I'd prefer it to not be enforced, and instead be configurable, so for now I think it is better if we don't change this.

> Adds prettier for formatting json,md,tsx,yml files.

For the markdown files we should add a formatting rule that limits the max-width of a line to 80. This is so that this is readable in 80x25 text mode.

Separately, we shouldn't assume the user uses `bun` (I don't), and we should go with the lowest common denominator which is `npm`.

> Instead of printing // it <description>, it will now print // It <description>. notice capitalized I and . at the end of it.

This is actually on purpose, the way you write the comment get's exactly copied to the output. If you were to write it capitalized and with ending dots, then it should be printed that way.

I think we should revert this change.

---

Everything else looks good to me. As a separate note, no need to change this PR at all, but in the future it would be nice for PRs to be self-contained. This PR has a few unrelated things in it, so it would be nice to have each get focused attention.

Again, thank you very much for taking the time 🙇 !

alexfertel commented on crates/bulloak/tests/check/Integration.t.sol:
Let's rename this file to lowercase.

Could you expand on why it was added?

smol-ninja replied:
never mind, I have removed it. It was a requirement for `solhint` checks but I have removed `solhint` as well since its something configured by the user and shouldn't be enforced by this tool.

smol-ninja commented:
@alexfertel I have reverted the changes as we discussed. Please let me know if the PR looks good now.

-------------------

> In this case, the codebase is the user of bulloak through the contributor 😆 Let me know if this makes more sense now

Yes it does and I agree that its not a good idea to format solidity files using configurations set in this repo. I have removed `forge fmt` and `solhint` integrations.

> I have had bad experiences with bun in the past

Sorry to hear that. My experience has been opposite and `bun` has been my go-to-choice for node packages for more than a year now. May be its time for you to try it again :))

Anyways I have removed `bun`.

> Yeah, this sounds good to me, I'm happy having this as a flag

Thanks. I have reverted these changes from this repo and created [an issue](https://github.com/alexfertel/bulloak/issues/108) to track it.

> I'm super happy that you decided to jump into it, and hopefully you like my code

💜💜💜

alexfertel reviewed (approved):
LGTM! Thank you very much!
